src/sage/features/threejs.py

```diff
@@ -56,8 +56,11 @@ def required_version(self):
 
         filename = os.path.join(SAGE_EXTCODE, 'threejs', 'threejs-version.txt')
 
-        with open(filename) as f:
-            return f.read().strip()
+        try:
+            with open(filename) as f:
+                return f.read().strip()
+        except FileNotFoundError:
+            return "unknown"
```

Comment:
This is overly complicated. Just kill the file and hardcode the string here. I'd even kill the whole "threejs version" business which is completely nuts and anti-modular. (Could all of this be replaced by https://github.com/jupyter-widgets/pythreejs ?)

Anyway, the whole modularization business is just too complicated. What do you mean "sage.features is shipped by sagemath-environment, but the version file that the Feature uses is shipped by sagemath-repl." Certainly all the pieces needed to use threejs (including the feature, etc) should go in the same "module".

Modularization should be about reducing complexity, not increasing it. Modules should have clear separation of concerns and communicate via clear well-defined interfaces.

Reply:
> the whole modularization business is just too complicated.

Your sentence is not complete. When you say "too ...", you need to say "for ..." as well.

Reply:
> > the whole modularization business is just too complicated.
> 
> Your sentence is not complete. When you say "too ...", you need to say "for ..." as well.

No, I don't. I think it's too complicated, period.

Reply:
> Modularization should be about reducing complexity, not increasing it.

You have already made it sufficiently clear that you're not interested, and that's fine; but I'll explain it nevertheless for the audience.

No, modularization is not primarily or absolutely about "reducing complexity". There are other criteria such as reusability and portability and deployability. Generally we decide to break out modules from a larger unit for a combination of such reasons. In the case of the Sage modularization project (#29705), "complexity" ranks last among these 4 criteria in importance.

Reply:
> What do you mean "sage.features is shipped by sagemath-environment, but the version file that the Feature uses is shipped by sagemath-repl." 

As you can see in 
- https://github.com/sagemath/sage/blob/develop/pkgs/sagemath-environment/pyproject.toml.m4#L42,

the Feature definitions are provided by the small distribution **sagemath-environment**, which by design has no dependencies -- it is for *detecting*, not *providing* features. That's the well-defined interface, separation of concerns etc. 

**sagemath-environment** certainly should not ship the plotting support code that we have in `sage/ext_data/threejs/` -- where the threejs-version.txt file resides.

Reply:
> **sagemath-environment** certainly should not ship the plotting support code that we have in `sage/ext_data/threejs/` -- where the threejs-version.txt file resides.

The declaration of the threejs feature belongs to the module using threejs. That is separation of concerns. It's none of the business of **sagemath-enviromnent** to know about the requirements of the plotting support code.

Reply:
> We agree on this; that's exactly the modularization regression that you introduced in #37024.

No, we don't agree. I did not introduce any regression. At most I surfaced a design bug with your "modularization" project. I never meant for sage.features.threejs to be shipped in sagemath-environment.

Reply:
> I did not introduce any regression.

You did, but no worries, I'm not making you pay a fine for it.

A design principle of `sage.features` is that client code can import any `Feature` defined there and then call it to test whether it's present. The feature definitions are centralized in `sage.features`; this reduces the complexity for the client code, as they can import from there without having to chase down the feature definition in the whole Sage library, or catch errors already when trying to import the feature definition. 

The design decision of shipping all of `sage.features` in one distribution **sagemath-environment** follows the same idea: Client code does not need to chase down which distribution to declare as a dependency; it's always just **sagemath-environment**.

Reply:
> with your "modularization" project

However, there *may* be a fine for using ironic quotation marks.

Reply:
> A design principle of `sage.features` is that client code can import any `Feature` defined there and then call it to test whether it's present. The feature definitions are centralized in `sage.features`; this reduces the complexity for the client code, as they can import from there without having to chase down the feature definition in the whole Sage library, or catch errors already when trying to import the feature definition.

Ok, I'm happy to switch to constructive mode. Let's try to think out of the box. You want "client code" to access a feature by name, without fiddling with import paths, etc.

The current way is:
```
from sage.features.threejs import Threejs
...
Threejs().is_present()
```
so you already need to know that the feature is `sage.featuers.threejs.Threejs` (.
But another feature is `sage.features.databases.DatabaseCremona` (not to be confused with `sage.databases.cremona.CremonaDatabase`), etc. And the feature name is `threejs` for the first one and `database_cremona_ellcurve`.. or maybe `database_cremona_mini_ellcurve`, and for the second you need to know to instantiate as `DatabaseCremona('cremona_mini')`.

On the other hand, there is already a registry of features which allows one to get the feature by name, although it's not very convenient:
```
sage: from sage.features.all import all_features
sage: [f for f in all_features() if f.name == 'threejs']
[Feature('threejs': JavaScript library to display 3D graphics)]
```

Wouldn't it make more sense to have a function to get a feature by name, and maybe another function to test if a feature is present by name? Of course, if the module containing the feature is not available, the feature is not available.

As for the registry itself: at first sight it looks like the kind of thing python entry points are designed for. In this way, a module could register, say, a `sage.features` entry point which maps to the feature.

> The design decision of shipping all of `sage.features` in one distribution **sagemath-environment** follows the same idea: Client code does not need to chase down which distribution to declare as a dependency; it's always just **sagemath-environment**.

Sure: place the sage.features support code in that distribution. But move the actual feature declaration to the distribution containing the feature.


> > with your "modularization" project
> 
> However, there _may_ be a fine for using ironic quotation marks.

No irony intended. Making clear that I refer to what you call "modularization" which is rather different of what I'd call "modularization", and based on a quick look of the initial responses your project got, I'm not alone.

Reply:
In any case: can you point out what code outside plotting actually needs to import the threejs feature? What would go wrong if the threejs feature is shipped in the same module as the actual plot code? Given that your design allows the different packages/distributions to be split at arbitrary file boundaries, the file could still be `sage.features.threejs`, but just shipped by a different distribution.

In fact, nothing really requires that threejs is a feature, all we really need is a function defined somewhere in `sage.repl` that will discover and cache the location of `three.js` from a reasonable set of possible locations. So now I'm actually regretting that I implemented this as a feature (same for databases).

Reply:
What I am finding valuable in the current design, with first-class `Feature` objects, is that you can get the documentation of a feature even when the feature is not loaded. This documentation includes the information from which distribution on PyPI or which distro package it is available etc.

This discoverability would be lost if the advertised interface is just "query by feature name", giving no information when the feature is not present.

Reply:
> As for the registry itself: at first sight it looks like the kind of thing python entry points are designed for.

Yes, I did consider this briefly when I wrote the `all_features` thing. But ultimately I think it's not the right tool for it. 

The Sage documentation will not be modularized: That we have one reference manual, not a loose collection of a hundred readthedocs sites, is a feature, not a bug. As the feature names are for controlling the examples in the Sage documentation, also the feature names are a global thing.

Reply:
> What I am finding valuable in the current design, with first-class `Feature` objects, is that you can get the documentation of a feature even when the feature is not loaded. This documentation includes the information from which distribution on PyPI or which distro package it is available etc.
> 
> This discoverability would be lost if the advertised interface is just "query by feature name", giving no information when the feature is not present.

We could have a "global" registry, in the form of a configuration file belonging to the sage.features module, which just lists the meta-information, i.e. name of the feature, module/distribution in which is defined, etc. The actual code to check the feature presence *and* how to search for the feature in the system (when applicable) can still belong to the module.

Reply:
> > As for the registry itself: at first sight it looks like the kind of thing python entry points are designed for.
> 
> Yes, I did consider this briefly when I wrote the `all_features` thing. But ultimately I think it's not the right tool for it.

Why not?

Reply:
I'm generally skeptical when using a "configuration file" is brought in as a proposed solution...

> The actual code to check the feature presence _and_ how to search for the feature in the system (when applicable) can still belong to the module.

That is already possible and does not require a redesign of the whole thing. See for example: https://github.com/sagemath/sage/blob/develop/src/sage/features/mip_backends.py#L33

Reply:
> Given that your design allows the different packages/distributions to be split at arbitrary file boundaries, the file could still be `sage.features.threejs`, but just shipped by a different distribution.

I thought I explained this already: If the module `sage.features.threejs` were shipped by a different distribution, then client code would have to already use `try: ... except:` when importing the feature.

Reply:
> I thought I explained this already: If the module `sage.features.threejs` were shipped by a different distribution, then client code would have to already use `try: ... except:` when importing the feature.

1. There's no "client code" for threejs outside of `sage.repl`, so there's no harm shipping `sage.features.threejs` together with `sage.repl`.
2. Even if there was client code, it could get the feature with something like `get_feature("threejs")`, so the `try: import ...` would be abstracted away.

It seems to me we are back to non-constructive mode.

Reply:
In the current implementation, `sage.features.threejs` gets the version info from a text file shipped as package data shipped by the responsible distribution. My fix just repairs this for the case that that distribution is not instsalled.

This could as well be replaced along lines similar to what you mention in https://github.com/sagemath/sage/pull/37178/files#discussion_r1470471049: `try: except` for importing a function that returns the version info. 

Either one is fine with me. Reworking the whole of the design of sage.features does not seem interesting to me